pallets/gear/src/lib.rs

```diff
@@ -1727,6 +1738,102 @@ pub mod pallet {
+        pub fn send_message_with_voucher(
+            origin: OriginFor<T>,
+            destination: ProgramId,
+            payload: Vec<u8>,
+            gas_limit: u64,
+            value: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let payload = payload
+                .try_into()
+                .map_err(|err: PayloadSizeError| DispatchError::Other(err.into()))?;
+            let who = ensure_signed(origin)?;
+            let origin = who.clone().into_origin();
+
+            Self::check_gas_limit_and_value(gas_limit, value)?;
+
+            let message = HandleMessage::from_packet(
+                Self::next_message_id(origin),
+                HandlePacket::new_with_gas(
+                    destination,
+                    payload,
+                    gas_limit,
+                    value.unique_saturated_into(),
+                ),
+            );
+
+            ensure!(
+                Self::program_exists(destination) && Self::is_active(destination),
+                Error::<T>::InactiveProgram
+            );
+
+            // Message is not guaranteed to be executed, that's why value is not immediately
+            // transferred. That's because destination can fail to be initialized by the time
+            // this dispatch message is next in the queue.
+            //
+            // Note: reservaton is made from the user's account as voucher can only be used
+            // to pay for gas or settle transaction fees, but not as source for value transfer.
+            CurrencyOf::<T>::reserve(&who, value.unique_saturated_into())
+                .map_err(|_| Error::<T>::InsufficientBalanceForReserve)?;
+
+            let gas_limit_reserve = T::GasPrice::gas_price(gas_limit);
+
+            // We want to charge the expenses on the voucher issued for the tx sender
+            // First we reserve enough funds on the account to pay for `gas_limit`.
+            // Note: this time around the cost of gas is covered by the voucher.
+            let voucher_id =
+                VoucherOf::<T>::redeem_with_id(who.clone(), destination, gas_limit_reserve)
+                    .map_err(|_| {
+                        log::debug!(
+                            "Failed to redeem voucher for user {:?} and program {:?}",
+                            who,
+                            destination,
+                        );
+                        Error::<T>::FailureRedeemingVoucher
+                    })?;
+
+            // # Safety
+            //
+            // The `unreachable` clause is due to the `message_id` being newly generated
+            // with `Self::next_message_id`.
+            //
+            // Note: using the `voucher_id` as the external parent of the gas node in order for
+            // the leftover being refunded back to the voucher account and not the user's account.
+            GasHandlerOf::<T>::create(voucher_id, message.id(), gas_limit)
```

Comment:
Self::create

Reply:
You can wrap all this functionality into one function and reuse for both extrinsics due to the fact that it differs only with account id owner of gas node